### PR TITLE
add std.conv.hexString() to replace core language hex strings

### PR DESCRIPTION
**Pull request by WalterBright**
This is a reboot of https://github.com/D-Programming-Language/phobos/pull/3058 written by @BBasile which was mysteriously closed.

Forum discussion: http://forum.dlang.org/thread/me4njq%241cul%241@digitalmars.com

### CODE REVIEW

**Comment by JakobOvrum** [comment]
I mentioned a number of issues in the other PR that Github now classifies as "outdated diffs" but the issues were not actually addressed. Can you have a look at them?

**Comment by WalterBright** [comment]
Links to which specific ones?

**Comment by JakobOvrum** [comment]
* `static` is redundant for module-level functions
 * s/`hexStrImpl!(C)(hexData)`/`hexStrImpl!C(hexData)`/ etc
 * user-facing compile-time interface should be value templates, not function templates:
```d
template hexString(string hexData, C = char)
    if (hexData.isHexLiteral && isSomeChar!C)
{
    immutable hexString = hexStrImpl!C(hexData);
}
```

------------

My main point was that I don't think we should expose the runtime functions as they are proposed by this patch, as they're not written as range algorithms.

File: std/conv.d [hunk 1]
```diff
@@ -5334,3 +5334,249 @@ unittest
         auto y = castFrom!(long*).to!int(x);
     }
 }
+
+/**
+The function $(D isHexLiteral) checks the consistency of a string for the
+functions $(D hexString) and $(D hexBytes).
+The result is only true if the input string is composed of ASCII white
```
Comment (Panke) [inline std/conv.d]: nitpick:  only true if ==> true if and only if

File: std/conv.d [hunk 2]
```diff
@@ -5334,3 +5334,249 @@ unittest
...
+    bool isH;
+    foreach(c; hexData)
+    {
+        isH = c.isHexDigit;
+        if ((!isH) & (!c.isWhite))
+            return false;
+        i += isH;
+    }
+    return !(i & 1) & (i > 0);
+}
```
Comment (Panke) [inline std/conv.d]: Use logical operators for logical operations and bit operators for bit operations.

File: std/conv.d [hunk 3]
```diff
@@ -5334,3 +5334,249 @@ unittest
...
+
+///
+unittest
+{
+    // test all the hex digits
+    static assert( ("0123456789abcdefABCDEF").isHexLiteral);
+    // empty or white strings are not valid
+    static assert( !"\r\n\t".isHexLiteral);
+    // but are accepted if the count of hex digits is even
```
Comment (Panke) [inline std/conv.d]: Why? At least mention that in the docs.

**Comment by JakobOvrum** [comment]
Ping @WalterBright

**Comment by WalterBright** [comment]
rebased

**Comment by WalterBright** [comment]
Revamped it according to comments here. Removed hexBytes. Adjusted semantics to exactly match hex strings in the core language, except that nelSep is also recognized as whitespace.

This is meant to replace core hex strings, not as something more. Something more may come about in the future as usage leads us.

**Comment by JakobOvrum** [comment]
Nice! LGTM.

File: std/conv.d [hunk 4]
```diff
@@ -5350,3 +5350,195 @@ unittest
...
+    }
+    return !(i & 1);
+}
+
+///
+unittest
+{
+    // test all the hex digits
+    static assert( ("0123456789abcdefABCDEF").isHexLiteral);
+    // empty or white strings are not valid
```
Comment (kuettler) [inline std/conv.d]: Is there one not too many in the comment?

File: std/conv.d [hunk 5]
```diff
@@ -5350,3 +5350,195 @@ unittest
...
+    static assert( "\r\n\t".isHexLiteral);
+    // but are accepted if the count of hex digits is even
```
Comment (kuettler) [inline std/conv.d]: Again, the comment is misleading

**Comment by kuettler** [comment]
LGTM

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Pull request merged**